README.md
# Restla

## Introduction

Restla is a full-stack Node.js web framework written in [TypeScript](http://typescript.org/). It is specifically
designed for developing the backend REST APIs of web applications.

Rather than reinvent the wheel, Restla provides a solid foundation that simplifies and integrates existing libraries
commonly used in Node.js development. Restla integrates the [Koa](http://koajs.com/) web framework and
[Squell](https://github.com/creativecuriositystudio/squell), a type-safe wrapper for the [Sequelize](http://docs.sequelizejs.com/en/latest/)
SQL ORM, to provide a completely promise-driven API that supports the async/await paradigm.

Restla extends Koa 2.x with additional functionality, but Koa's core functionality remains the same.
This means you can use any official or third-party Koa middleware with Restla. By default, Restla applications
have the following Koa middleware enabled:

* [koa-bodyparser 3.x](https://github.com/koajs/bodyparser)
* [koa-router 7.x](https://github.com/koajs/bodyparser)

## Features

* All Koa middleware works out of the box with Restla applications.
* `Resource`, a router that can generate generic REST resource routes from a Squell model.
  The default functionality can easily be change by extending the `Resource` class.
* `Auth`, an authentication helper that is backend agnostic (i.e. you could authenticate
  with a third-party authentication.

## Installation

```
npm install --save restla
```

## Usage

### Error Handling

By default Restla catches all error during requests, coerces them
into `ApplicationError`s if they're not already an application error
and then sends them to the client with a response similar to:

```json
{
  "message": "Some error message",
  "errors": []
}
```

Restla automatically coerces ModelSafe validation errors into a 400 response
and any authentication errors into 402 responses. Any other unknown errors
are then turned into 500 errors. The validation error response (400 bad request)
looks similar to the above but is populated with error messages for each field:

```json
{
  "message": "Validation failed",
  "errors": [{
    "path": "name",
    "message": "Is required"
  }]
}
```

If you hit any errors you should reject (or throw if you're using the async keyword) with an an `ApplicationError`
in your route or resource. It takes a status code and error message like so:

```typescript
throw new ApplicationError(404, 'Not Found');
```

Restla will automatically catch any rejected errors and send them using the `ApplicationContext`'s error method.
You can provide your own error response handling method by passing in a custom `ApplicationContext` when instantiating
a Restla application.

## Documentation

The API documentation generated using [TypeDoc](https://github.com/TypeStrong/typedoc)
is [available online](http://creativecuriosity.github.io/restla).

To generate API documentation from the code into the `docs` directory, run:

```sh
npm run docs
```

## Testing

First install the library dependencies and the SQLite3 library:

```
npm install
npm install sqlite3
```

To execute the test suite using SQLite as the backend, run:

```
npm run test
```


CHANGELOG.md
# 0.8.3

* Added authentication to resource auth handling

# 0.8.2

* Use Bluebird promise for Acl

# 0.8.1

* Fixed resource milestone bug

# 0.8.0

* Added ability to specify public actions in the resource options

# 0.7.5

* Provide better error tracing when coercing an error

# 0.7.4

* Give an error message when throwing an `AuthorisationError` so it's clear what is going on

# 0.7.3

* Something...

# 0.7.2

* Something...

# 0.7.1

* Bump to fix broken build (missing index type map)

# 0.7.0

* Integrate ACL into `Auth` to allow for authorisation of routes and resources
* Introduce new `auth` resource milestone where authorisation is performed

# 0.6.1

* Simplify association includes, and set 'associateOnly' so assocs' data isn't saved when the parent is updated
* Upgrade squell and modelsafe

# 0.6.0

* Upgrade to squell & modelsafe `1.0.0-alpha`

# 0.5.0

* Upgrade to Squell & ModelSafe `0.7.0`
* Improve error stack generated when coercing to `ApplicationError` to include the old stack

# 0.4.3

* Fix delete route so it actually works

# 0.4.2

* Fix association option not being considered on list actions

# 0.4.1

* Fix multipart incorrectly skipping multiparts

# 0.4.0

* Bump Squell and ModelSafe dependencies for bug fixes
* Add multipart parsing option (using formidable)
* Change responder class methods to have the router context as `this`

# 0.3.2

* Fix token production not awaiting
* Fix usage of moment.js import

# 0.3.1

* Fix responder export

# 0.3.0

* Remove application context, modifying REST responses is now done through the `Responder` class
* Manual application error coercing has been converted to a handle based system,
  allowing users of the library to register their own handlers for their own error types
  that coerce to an application error

# 0.2.2

* Bump Squell required to 0.5.2 for association saving fix

# 0.2.1

* Broke out error coercion into `ApplicationError.coerce`

# 0.2.0

* Correct error handling, and have all error types correctly support the instanceof operator
* Move to Squell & ModelSafe 0.5
* Add `associations` flag for including all associations in a resource body

# 0.1.0

* Initial release


src/app.ts
/**
 * Provides the core features of a Crea application.
 */
import * as KoaApplication from 'koa';
import * as bodyParser from 'koa-bodyparser';
import * as _ from 'lodash';
import { Fields, Files, IncomingForm } from 'formidable';
import { Model, ValidationError } from 'modelsafe';

import * as auth from './auth';
import { Responder, ResponderConstructor } from './responder';
import { RouterContext } from './router';

/** An error in an application. */
export class ApplicationError extends Error {
  /** The HTTP status code of the error, to be sent in response. */
  status: number;

  /**
   * Any error data to be sent in response.
   * It is completely up to the instantiater as to what
   * to populate here - it
   */
  data?: any;

  /**
   * Construct an application error.
   *
   * @param status The HTTP status code.
   * @param message The error message.
   * @param data The error data.
   */
  constructor(status: number, message: string, data?: any) {
    super(message);

    this.name = 'ApplicationError';
    this.stack = new Error().stack;
    this.status = status;
    this.data = data;
  }

  /**
   * All of the registered handlers for coercing non-`ApplicationErrors` into
   * their relevant `ApplicationError` form.
   */
  // tslint:disable-next-line:ban-types
  static handlers: Map<Function, (err: any) => ApplicationError> = new Map();

  /**
   * Coerce any error type into an `ApplicationError`.
   *
   * This will lookup the registered coercion handlers based off the error provided.
   * If no coerce handler has been provided, then th
   *
   * By default the following coercion handlers are enabled:
   *
   * * modelsafe.ValidationError: 400 Bad Request, with the `data` field populated with the errors for each request field
   * * restla.TokenExpiryError: 401 Unauthorised
   * * restla.TokenInvalidError: 401 Unauthorised
   * * restla.UserNotFoundError: 401 Unauthorised
   * * restla.AuthorisationError: 403 Forbidden
   */
  static coerce(err: Error | ApplicationError): ApplicationError {
    if (err instanceof ApplicationError) {
      return err as ApplicationError;
    }

    // Lookup the registered handler by error constructor as the key.
    const handler = ApplicationError.handlers.get(err.constructor as typeof Error);

    // Call the handler or create a default error
    let result = typeof (handler) === 'function' ?
      handler(err as Error) :
      new ApplicationError(500, err.message);

    // Merge the stacks.
    result.stack = `${result.stack}\ncaused by ${err.stack}`;

    return result;
  }

  /**
   * Register a handler for coercing another error type
   * into an application error.
   *
   * This is useful if you ever intend to throw non-`ApplicationError`s
   * anywhere in your code - Restla will automatically catch those
   * and then look for a coerce handler for that error. If no handler
   * has been registered, then it will just coerc into a generic 500 status
   * `ApplicationError` with the error message carried across.
   *
   * @see coerce
   * @param ctor The error constructor to automatically coerce using the handler.
   * @param handler The handler that turns the relevant error type into an application error.
   */
  // tslint:disable-next-line:ban-types
  static register<T extends Error>(ctor: Function, handler: (err: T) => ApplicationError) {
    ApplicationError.handlers.set(ctor, handler);
  }
}

ApplicationError.register(ValidationError, <T extends Model>(err: ValidationError<T>): ApplicationError => {
  return new ApplicationError(400, err.message, err.errors);
});

ApplicationError.register(auth.TokenInvalidError, (err: auth.TokenInvalidError): ApplicationError => {
  return new ApplicationError(401, err.message);
});

ApplicationError.register(auth.TokenExpiryError, (err: auth.TokenExpiryError): ApplicationError => {
  return new ApplicationError(401, err.message);
});

ApplicationError.register(auth.UserNotFoundError, (err: auth.UserNotFoundError): ApplicationError => {
  return new ApplicationError(401, err.message);
});

ApplicationError.register(auth.AuthorisationError, (err: auth.AuthorisationError): ApplicationError => {
  return new ApplicationError(403, err.message);
});

/** Options for running an application. */
export interface ApplicationOptions {
  /** The responder class to use for sending REST responses to clients. */
  responder: ResponderConstructor;

  /**
   * Whether multipart requests should be parsed. Off by default to prevent overhead
   * in apps that don't need file uploads or other multipart requests.
   *
   * If a set of formidable options are provided instead of a boolean,
   * then they will be used as the options for parsing multipart with formidable.
   */
  multipart: boolean | Partial<IncomingForm>;
}

/**
 * The web application.
 *
 * This is functionally equivalent to a Koa application,
 * except we use some must-have/good-to-have middlewares
 * early on and provide some extra things, such as smart error handling.
 */
export class Application extends KoaApplication {
  /** The options for the application. */
  protected options: ApplicationOptions;

  /**
   * Construct an application.
   *
   * @param options Any options for the application.
   */
  constructor(options?: Partial<ApplicationOptions>) {
    super();

    this.options = options = {
      multipart: false,
      responder: Responder,

      ... options
    };

    // Setup default values on the router context, such as the responder.
    this.use(async (ctx: RouterContext, next: () => Promise<any>) => {
      ctx.responder = new options.responder(ctx) as Responder;

      try {
        return await next();
      } catch (err) {
        // Coerce the error then send the response.
        return ctx.responder.error(ApplicationError.coerce(err));
      }
    });

    if (options.multipart) {
      // Parse multipart forms using formidable, if multipart mode is on.
      this.use(async (ctx: RouterContext, next: () => Promise<any>) => {
        if (!ctx.request.is('multipart/form-data')) return next();

        let form = new IncomingForm();

        // If a non-boolean was provided, extend the form with the user's options.
        if (typeof (options.multipart) !== 'boolean') {
          _.extend(form, options.multipart as Partial<IncomingForm>);
        }

        [ctx.request.fields, ctx.request.files] = await new Promise(
          (resolve: (x: [Fields, Files]) => void, reject: (err: any) => void) => {
            form.parse(ctx.req, (err, fields, files) => {
              // Reject and our app error handler will pick it up
              if (err) {
                return reject(err);
              }

              resolve([fields, files]);
            });
          }
        );

        return next();
      });
    }

    // Add any default middleware.
    this.use(bodyParser());
  }
}
